File: exercises/js/day1-send-xrp.js
// ============================================================
// Day 1 Exercise: Send XRP
// Learn: How to construct and submit a Payment transaction
// on the XRP Ledger testnet.
// Reference materials:
// - https://learn.xrpl-commons.org/course/blockchain-foundations-for-web2-developers/lesson/create-accounts-and-send-xrp/
// - https://xrpl.org/docs/references/protocol/transactions/types/payment
// - https://js.xrpl.org/
// ============================================================

const xrpl = require('xrpl');
const fs = require('fs');
const { notifyProgress } = require('../../scripts/notify');

const TESTNET_URL = 'wss://s.altnet.rippletest.net:51233';

async function main() {
  const { wallets } = JSON.parse(fs.readFileSync('../../wallets.json', 'utf-8'));
  const senderWallet = xrpl.Wallet.fromSeed(wallets[0].seed);
  const receiverWallet = xrpl.Wallet.fromSeed(wallets[1].seed);

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();
  console.log('Connected to XRPL Testnet');

  // Check balance before
  const balanceBefore = await client.getXrpBalance(receiverWallet.address);
  console.log(`Receiver balance before: ${balanceBefore} XRP`);

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 1: Construct the Payment transaction object                ║
  // ║                                                                  ║
  // ║ HINT: {                                                          ║
  // ║   TransactionType: 'Payment',                                    ║
  // ║   Account: senderWallet.address,                                 ║
  // ║   Amount: xrpl.xrpToDrops('10'),                                 ║
  // ║   Destination: receiverWallet.address                             ║
  // ║ }                                                                ║
  // ╚══════════════════════════════════════════════════════════════════╝
  const paymentTx = {
    // YOUR CODE HERE
  };

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 2: Submit the transaction and log the result               ║
  // ║                                                                  ║
  // ║ HINT: const result = await client.submitAndWait(paymentTx,       ║
  // ║         { wallet: senderWallet });                                ║
  // ║       console.log(result);                                        ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  // Check balance after
  const balanceAfter = await client.getXrpBalance(receiverWallet.address);
  console.log(`Receiver balance after: ${balanceAfter} XRP`);

  console.log(`\nExplorer: https://testnet.xrpl.org/accounts/${senderWallet.address}`);
  await notifyProgress('Day 1: Send XRP');
  await client.disconnect();
}

main().catch(console.error);

File: exercises/python/day1_send_xrp.py
"""
Day 1 Exercise: Send XRP
========================
Send 10 XRP from wallet[0] to wallet[1] on XRPL Testnet.

Reference materials:
- https://learn.xrpl-commons.org/course/blockchain-foundations-for-web2-developers/lesson/create-accounts-and-send-xrp/
- https://xrpl.org/docs/references/protocol/transactions/types/payment
- https://xrpl-py.readthedocs.io/en/stable/
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import Payment
from xrpl.transaction import submit_and_wait
from xrpl.utils import xrp_to_drops
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'scripts'))
from notify_progress import notify_progress

TESTNET_URL = 'wss://s.altnet.rippletest.net:51233'

# Load wallets
with open('../../wallets.json') as f:
    data = json.load(f)
    wallets = data['wallets']

sender = Wallet.from_seed(wallets[0]['seed'])
receiver = Wallet.from_seed(wallets[1]['seed'])

print(f"Sender:   {sender.address}")
print(f"Receiver: {receiver.address}")

with WebsocketClient(TESTNET_URL) as client:

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 1: Create a Payment transaction                             ║
    # ║                                                                  ║
    # ║ HINT: Payment(account=sender.address,                            ║
    # ║              amount=xrp_to_drops(10),                            ║
    # ║              destination=receiver.address)                       ║
    # ╚══════════════════════════════════════════════════════════════════╝

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 2: Submit the transaction and print the result              ║
    # ║                                                                  ║
    # ║ HINT: result = submit_and_wait(payment_tx, client, sender)       ║
    # ║       print(result)                                              ║
    # ╚══════════════════════════════════════════════════════════════════╝
    pass

# Explorer: https://testnet.xrpl.org/accounts/{address}
print(f"\nCheck sender:   https://testnet.xrpl.org/accounts/{sender.address}")
print(f"Check receiver: https://testnet.xrpl.org/accounts/{receiver.address}")
notify_progress('Day 1: Send XRP')
